Handle registered meta and registered rest fields

eri-trabiccolo opened the pull request:
## Description

Fixes #157 
also fixes #73 

## How has this been tested?
manually

## Screenshots <!-- if applicable -->

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. <!-- Check code: `composer run-script tests-run`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/tests/README.md -->
- [x] My code follows the LifterLMS Coding & Documentation Standards. <!-- Check code: `composer run-script check-cs-errors`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/docs/coding-standards.md and https://github.com/gocodebox/lifterlms/blob/trunk/docs/documentation-standards.md -->

eri-trabiccolo committed:
Handle registered meta and registered rest fields

thomasplevy commented on includes/abstracts/class-llms-rest-controller.php:
```diff
@@ -754,4 +1010,32 @@ public function update_item( $request ) {
+
+		if ( ! empty( $schema['properties']['meta'] ) && isset( $request['meta'] ) ) {
+
+			$request = parse_custom_meta_fields( $request['meta'] );
```
```suggestion
			$request = $this->parse_custom_meta_fields( $request['meta'] );
```

thomasplevy reviewed (commented):
A few things I noticed while testing this (I only tried via courses endpoint)

1) If no `meta` prop is submitted during creation an error is returned. I think this is probably just an oversight.
2) If you pass in unregistered meta data during a creation the request is accepted without error (but it's not returned in the response). I don't know if this is the *proper* behavior but I noticed it... On further inspection it seems this is the same behavior present in the WP core so we're probably fine here.

The course should exclude the rest field registered by the block plugin `_llms_length` as a disallowed field.

Other than that... I could not figure out how to actually get any custom data to work. I registered a meta field that I could get to show up but I couldn't get any data passed into it properly (maybe this was a registration issue?)

I couldn't get a custom rest field to work either.

All that said... I think this is a solid strategy but I'm having a hard time signing off on it as an "okay way to proceed" since I can't really get it working. Maybe in lieu of tests you could provide the example code your using to test this manually so that I could use that to review in greater detail...

eri-trabiccolo commented:
@thomasplevy thanks.
Have to check the issue you reported.
I might have made some changes after the latest working test :) 
About `_llms_length`... well it should be skipped automatically because of `length` being an item schema property, if it doesn't happen, that's another issue to look into. Thank you man! ❤️

thomasplevy commented:
> About `_llms_length`... well it should be skipped automatically because of `length` being an item schema property, if it doesn't happen, that's another issue to look into. Thank you man! heart

I saw this when performing an `OPTIONS /courses` to review the schema when trying to add fields. I tried to exclude it too (via the disallowed_meta prop) and it was still showing up on the OPTIONS request. I didn't actually try to set it and it wasn't being returned in the meta array.

eri-trabiccolo committed:
Update includes/abstracts/class-llms-rest-controller.php

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo committed:
Fix meta fields not correctly excluded from the schema

eri-trabiccolo commented:
@thomasplevy you were right about the `_llms_length` I wasn't exlcuding it from the schema, only from the responses.
fixed in d55c312

eri-trabiccolo committed:
fix additional fields handling

eri-trabiccolo commented:
You were right about the registered fields (~via `register_rest_field` - haven't re-tested the ones via `register_meta` though, I'll do asap~ ) too, there was an error in how I handled the schema properties skipping.
I fixed it in ~c16d03d~75dcb26
I used this example to test the rest field:
https://developer.wordpress.org/reference/functions/register_rest_field/#comment-3434
just changing `post` with `course` and `$object->ID` with `$object->get('ID')`

Fixed meta update in 141cca0

eri-trabiccolo committed:
Additional fields fix

eri-trabiccolo committed:
fix meta update

eri-trabiccolo committed:
fix phpcs

eri-trabiccolo committed:
Additional fields fix - reprise

eri-trabiccolo committed:
clean up

eri-trabiccolo committed:
fix meta prefixing

eri-trabiccolo committed:
clean logic

eri-trabiccolo committed:
clean up

eri-trabiccolo committed:
update_additional_fields_for_object requires the created object as first param

eri-trabiccolo committed:
fix adding meta properties to the schema

eri-trabiccolo committed:
allow all resources to handle custom fields registered via register_rest_field

eri-trabiccolo committed:
docs

eri-trabiccolo committed:
Added tests on post type resources with meta

eri-trabiccolo committed:
Added tests on user resources with meta

eri-trabiccolo committed:
add users tests

eri-trabiccolo committed:
fix tests

eri-trabiccolo committed:
add some tests on custom rest fields handling

eri-trabiccolo committed:
fix webhooks not correctly handling additional fields

eri-trabiccolo committed:
some new tests refactoring to reduce redundancy

eri-trabiccolo committed:
fix enrollment failing test

eri-trabiccolo committed:
add tests

eri-trabiccolo committed:
tests improvements

eri-trabiccolo committed:
test updating registered fields

eri-trabiccolo committed:
fix webhooks creation/update when using custom rest fields

eri-trabiccolo committed:
improve tests for custom fields

eri-trabiccolo committed:
fix meta fields exclusion

eri-trabiccolo committed:
prepare_object_data_for_response can be protectd rather than public

eri-trabiccolo committed:
add changelog for schema caching

eri-trabiccolo commented:
This massive PR is ready to be reviewed.
Includes some code reorganization/refactoring (not exhaustive though) to reduce some redundancy.
What it basically does is:
- caching the item schema (see #73)
- Allow post based resources which support custom fields, like course, membership, lesson..., so not section, to manage (CRUD) post meta, as well as user based resources to manage user meta.
- Allow all the resources to be extended using `register_rest_field()`. Note that, differently from the WP Core, we don't allow `register_rest_field()` to override a property already defined in the schema. E.g. Take the course's `access_opens_date` which is defined by our schema, and for which we have specific creation/update code, you won't be able to register a field called `access_opens_date`.

eri-trabiccolo committed:
changelogs

eri-trabiccolo committed:
Don't call `rest_ensure_response()` twice, already called in `$this->prepare_item_for_response()`

thomasplevy commented:
@eri-trabiccolo sorry I haven't gotten a look at this yet. We'll get to it thank you <3

eri-trabiccolo commented:
@seothemes 
can I get your functional review on this? Test everything works fine.
Let me know when you can start and we'll chat a bit on the aims of this PR if it's not clear (it's a pretty big one :D)

seothemes reviewed (approved):
This is all good functionality wise 👍

eri-trabiccolo commented:
Thanks!
Gonna rebase and merge! <3

eri-trabiccolo committed:
Update issue_157-new-1.yml

eri-trabiccolo committed:
Update issue_157-new-2.yml

eri-trabiccolo committed:
fix typos
